File: problems/longest_substring_without_repeating_characters/longest_substring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        # One pass, no look back
        # q p x r j x k l t z y q
        # ^ ------- ^
        #       ^ ------------- ^
        # x ................... ^ (don't count since start > chars[char] + 1)
        #       start           i
        longest = start = 0
        chars = dict()
        for i, char in enumerate(s):
            if char in chars:
                # Stop counting and check how long the current length is.
                longest = max(longest, i - start)  # Not include the current.

                # Don't upgrade the start if it is smaller than the repeat char
                # Otherwise, you will include a duplicated char in the middle.
                start = max(start, chars[char] + 1)

            # Record the last possition for each character.
            chars[char] = i
        return max(longest, len(s) - start)

if __name__ == "__main__":
    print Solution().lengthOfLongestSubstring("dvdf")

File: problems/longest-substring-without-repeating-characters/longest_substring_without_repeating_characters.cc
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    int max_length = 0;
    int p = 0;  // The start of the candidate string.
    int q = 0;  // The end of the candidate string.
    unordered_map<char, bool> exist;
    while (q < s.length()) {
      if (exist[s[q]]) {
        max_length = max(max_length, q - p);
        while (s[p] != s[q]) {  // Find the repeat character.
          exist[s[p]] = false;
          p++;
        }
        p++;
        q++;
      } else {
        exist[s[q]] = true;
        q++;
      }
    }
    max_length = max(max_length, q - p);
    return max_length;
  }
};

int main(int argc, char** argv) {
  Solution solution;
  cout << solution.lengthOfLongestSubstring("c") << endl;
  return 0;
}